cpp/src/streaming/core/spillable_messages.cpp

```diff
@@ -0,0 +1,78 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <rapidsmpf/streaming/core/spillable_messages.hpp>
+
+namespace rapidsmpf::streaming {
+
+SpillableMessages::MessageId SpillableMessages::insert(Message&& message) {
+    std::lock_guard<std::mutex> lock(global_mutex_);
+    content_descriptions_.insert({counter_, message.content_description()});
+    items_.insert({counter_, std::make_shared<Item>(std::move(message))});
+    return counter_++;
+}
+
+Message SpillableMessages::extract(MessageId mid) {
+    std::unique_lock global_lock(global_mutex_);
+    std::shared_ptr<Item> item = extract_item(items_, mid).second;
+    content_descriptions_.erase(mid);
+    global_lock.unlock();
+
+    // If the item is being spilled, we block here until the spilling is done.
+    std::unique_lock item_lock(item->mutex);
+    RAPIDSMPF_EXPECTS(item->message.has_value(), "message not found", std::out_of_range);
```

Comment:
question: How can this occur? Noone else can have a reference to `item` at this point, right?